cloud-storage/googledrive/impl/src/java/org/sakaiproject/googledrive/service/GoogleDriveServiceImpl.java

```diff
@@ -0,0 +1,359 @@
+		}
+		return false;
+	}
+
+	public Drive refreshToken(String userId){
+		if(!isConfigured()){
+			return null;
+		}
+		Drive googledriveUser = googledriveUserCache.get(userId);
+		if(googledriveUser != null) {
+			log.debug("refreshToken : Reusing previous user data " + googledriveUser);
+			return googledriveUser;
+		}
+
+		try {
+			Credential storedCredential = flow.loadCredential(userId);
+			if (storedCredential == null) {
+				log.warn("Couldn't retrieve Google credential for user {}", userId);
+				return null;
+			}
+
+			// This gives us is the ability to update our stored
+			// credentials as they get refreshed (using the
+			// DataStoreCredentialRefreshListener).
+			GoogleCredential credential = new GoogleCredential.Builder()
+					.setTransport(httpTransport)
+					.setJsonFactory(jsonFactory)
+					.setClientSecrets(clientSecrets)
+					.addRefreshListener(new CredentialRefreshListener() {
+						public void onTokenErrorResponse(Credential credential, TokenErrorResponse tokenErrorResponse) {
+							log.error("OAuth token refresh error: {}", tokenErrorResponse);
+						}
+
+						public void onTokenResponse(Credential credential, TokenResponse tokenResponse) {
+							log.debug("OAuth token was refreshed");
+						}
+					})
+					.addRefreshListener(new DataStoreCredentialRefreshListener(userId, flow.getCredentialDataStore()))
+					.build();
+			credential.setRefreshToken(storedCredential.getRefreshToken());
+			credential.setAccessToken(storedCredential.getAccessToken());
+			log.debug("credential.getServiceAccountUser() : {}", credential.getServiceAccountUser());
+
+			service = new Drive.Builder(httpTransport, jsonFactory, credential)
+				.setApplicationName(GOOGLEDRIVE_APP_NAME)
+				.build();
+
+			googledriveUserCache.put(userId, service);
+			return service;
+		} catch(Exception e) {
+			log.warn("GoogleDrive: Error while refreshing or saving the token for user {} : {}", userId, e.getMessage());
+		}
+		return null;
+	}
+
+	public GoogleDriveUser getGoogleDriveUser(String userId) {
+		return googledriveRepo.findBySakaiId(userId);
+	}
+
+	public List<GoogleDriveItem> getDriveRootItems(String userId) {
+		if(!isConfigured()){
+			return null;
+		}
+		List<GoogleDriveItem> cachedItems = driveRootItemsCache.get(userId);
+		if(cachedItems != null) {
+			log.debug("getDriveRootItems : Returning cached items " + cachedItems);
+			return cachedItems;
+		}
+		List<GoogleDriveItem> items = new ArrayList<>();
+		try {
+			service = refreshToken(userId);
+			if(service == null){
+				return null;
+			}
+
+			Drive.Files.List list = service.files().list();
+			list.setFields(String.format("nextPageToken, files(%s)", String.join(",", fileFieldsToRequest)));
+			list.setOrderBy("folder,name");
+			String queryString = "'root' in parents";
+			list.setQ(queryString);
+			FileList result = list.execute();
+			List<File> files = result.getFiles();
+			for (File file : files) {
+				log.debug("Google data : {}", file);
```

Comment:
You should guard this.

Reply:
And this one doesn't need to be guarded.